sage/api/src/traits.rs

```diff
@@ -19,17 +19,28 @@ pub trait SageGameTransition {
 	type TransitionId: Member + Parameter + MaxEncodedLen + TypeInfo;
 	/// THe type of the specific configuration options for the transition
 	type TransitionConfig: Member + Parameter + MaxEncodedLen + TypeInfo + Default;
+	/// The account id type, usually 32 bytes long.
 	type AccountId: Member + Codec;
+	/// The asset id type of the sage assets defined by the developer.
 	type AssetId: Member + Parameter + MaxEncodedLen + TypeInfo;
+	/// The asset type used in sage defined by the game developer.
 	type Asset: Member + Parameter + MaxEncodedLen + TypeInfo + GetId<Self::AssetId>;
 	/// An optional extra, which is simply forwarded to the `verify_rule` and `do_transition`
 	/// method. If you don't need custom arguments, you can define that type as `()`.
 	type Extra: Member + Parameter + MaxEncodedLen + TypeInfo + Default;
 
+	/// Defines the fungible asset that was used to pay the transaction. If the transition accesses
+	/// user funds, it might want to use the same asset, as this implies that the user is willing
+	/// to use this one instead of the native balance.
+	///
+	/// If you don't need fungible tokens in your transition you can set this as ().
+	type PaymentFungible;
+
 	fn do_transition(
 		transition_id: &Self::TransitionId,
 		account_id: &Self::AccountId,
 		assets_ids: &[Self::AssetId],
-		extra: &Self::Extra,
+		extra: Option<Self::Extra>,
```

Comment:
I don't think forcing the `extra` argument to be an `Option` is a good idea, since you can already define that in your implementation if needed. 

Also, if you don't need the extra at all you can just define it as the empty type and be done with it.

Reply:
I can see your point, but what if only one of your transitions needs an extra, and all others don't need it? I see that your point is that the extra type itself can be an option, but I do think that this is a bit more obscure and harder to understand.

I will think a bit more about it to see if there are actual hard arguments for either option.

Reply:
yeah, the thing is that fundamentally the extra argument can be part of the transition_id entirely anyways, so maybe the best approach is to simplify the trait by removing it

Reply:
I was thinking of the design you mentioned before, but then I think it is wrong to separate the transition and the id. There should be no more ID then, just a rich rust enum:

```rust
pub enum HeroTransition {
    Work { asset: ..., other_args }
    ...
}

impl HeroTransition {
    execute(self) -> Result<TransitionResult, TransitionError> {
        match self {
           Work { .... } => // do work      
        }
    }
}
```

This approach would already implicitly check 50% of the rules we have, and make them obsolete, and the only downside is the relatively bad rust enum support in other programming languages.

Reply:
hmmm, I keep coming back to this idea, and I start to think that bad enum support in other languages is not enough reason to not do it in rust.